### PR TITLE
complete tests and rest

### PR DESCRIPTION
**Pull request by DanyloPlotskykh**

### CODE REVIEW

File: _server/src/BusinessLogic/UsersModule.cpp [hunk 1]
```diff
@@ -38,6 +38,22 @@ ResponseData UsersModule::executeTask(const RequestData &requestData)
     {
         response = registerUser(requestData.dataset);
     }
+    else if (requestData.submodule == "roles" && requestData.method == "GET")
```
Comment (VolodymyrBilenko) [inline _server/src/BusinessLogic/UsersModule.cpp]: You perform the same comparison (**requestData.submodule == "roles"**) several times. It shall be done only once. Also you could use switch case for **requestData.method**, but it is up to you.

File: _server/src/BusinessLogic/UsersModule.cpp [hunk 2]
```diff
@@ -123,3 +139,92 @@ ResponseData UsersModule::registerUser(const Dataset &request)
...
+            response.dataset[Keys::Role::ID].emplace_back(i->id);
+            response.dataset[Keys::Role::NAME].emplace_back(i->name);
+        }
+    }
+    else
+    {
+        response.dataset[Keys::Role::ID] = {""};
+        response.dataset[Keys::Role::NAME] = {""};
+        SPDLOG_WARN("UsersModule::getRoles | array of users are empty");
+        return response;
```
Comment (VolodymyrBilenko) [inline _server/src/BusinessLogic/UsersModule.cpp]: One return statement at the end of the function will be enough

File: _server/src/BusinessLogic/UsersModule.cpp [hunk 3]
```diff
@@ -123,3 +139,92 @@ ResponseData UsersModule::registerUser(const Dataset &request)
...
     return response;
 }
+
+ResponseData UsersModule::getRoles()
+{
+    SPDLOG_TRACE("UsersModule::getRole");
+    ResponseData response;
+    auto rolesVec = m_rolesRepository->getAll();
+
+    if (!rolesVec.empty())
```
Comment (VolodymyrBilenko) [inline _server/src/BusinessLogic/UsersModule.cpp]: comparison approach differs from other lines (181,203). Decide which style to adhere and follow it everywhere

File: _server/src/BusinessLogic/UsersModule.cpp [hunk 4]
```diff
@@ -123,3 +139,92 @@ ResponseData UsersModule::registerUser(const Dataset &request)
...
+        {
+            response.dataset[Keys::Role::ID].emplace_back(i->id);
+            response.dataset[Keys::Role::NAME].emplace_back(i->name);
+        }
+    }
+    else
+    {
+        response.dataset[Keys::Role::ID] = {""};
+        response.dataset[Keys::Role::NAME] = {""};
+        SPDLOG_WARN("UsersModule::getRoles | array of users are empty");
```
Comment (VolodymyrBilenko) [inline _server/src/BusinessLogic/UsersModule.cpp]: array of users is empty

File: _server/src/BusinessLogic/UsersModule.cpp [hunk 5]
```diff
@@ -123,3 +139,92 @@ ResponseData UsersModule::registerUser(const Dataset &request)
...
+}
+
+ResponseData UsersModule::updateRoles(const Dataset &request, const std::string &resourseId)
+{
+    SPDLOG_TRACE("UsersModule::updateRoles");
+    if (resourseId.empty())
+    {
+        throw ServerException(_M, "resourseId is emtpy");
+    }
+    ResponseData response;
```
Comment (VolodymyrBilenko) [inline _server/src/BusinessLogic/UsersModule.cpp]: What is the purpose of this variable? You don't process it. Moreover you return it.

File: _server/src/BusinessLogic/UsersModule.cpp [hunk 6]
```diff
@@ -123,3 +139,92 @@ ResponseData UsersModule::registerUser(const Dataset &request)
...
+    {
+        SPDLOG_WARN("UsersModule::addNewRole username or password does not exit");
+        throw ServerException(_M, "Server got no username or password in Dataset");
+    }
+    auto usersVec = m_usersRepository->getByField(Keys::Role::NAME, roleName);
+    if (usersVec.empty())
+    {
+        m_rolesRepository->add(Role("", roleName));
+        SPDLOG_INFO("UsersModule::registerUser | new role `{}` is add", roleName);
+        response.dataset[Keys::Role::NAME] = {roleName};
```
Comment (VolodymyrBilenko) [inline _server/src/BusinessLogic/UsersModule.cpp:190]: Occasionally you use at(), sometimes []. Please use the same approach everywhere

File: _server/src/BusinessLogic/UsersModule.cpp [hunk 7]
```diff
@@ -123,3 +139,92 @@ ResponseData UsersModule::registerUser(const Dataset &request)
...
+    catch (std::out_of_range &e)
+    {
+        SPDLOG_WARN("UsersModule::updateRole role name does not exit");
+        throw ServerException(_M, "Server got no role name in Dataset");
+    }
+    m_rolesRepository->update(Role(resourseId, roleName));
+    return response;
+}
+
+ResponseData UsersModule::deleteRole(const std::string &resourseId)
```
Comment (VolodymyrBilenko) [inline _server/src/BusinessLogic/UsersModule.cpp:223]: Can't we return this structure as a reference?

File: _server/src/BusinessLogic/UsersModule.hpp [hunk 8]
```diff
@@ -24,6 +24,11 @@ class UsersModule : public IBusinessModule
     ResponseData loginUser(const Dataset &request);
     ResponseData registerUser(const Dataset &request);
 
+    ResponseData getRoles();
```
Comment (VolodymyrBilenko) [inline _server/src/BusinessLogic/UsersModule.hpp:27]: **void** as an argument shall be added

File: _server/tests/unit/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,30 @@
+# ------------------------- UNIT -------------------------
+set(SERVER_UNIT_TESTS_DIR
```
Comment (VolodymyrBilenko) [inline _server/tests/unit/CMakeLists.txt]: There is no sense for creating this variable. You can add sources directly to the add_executable

File: _server/tests/unit/UserModuleTest.cpp [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+        .module = "users",
+        .submodule = "roles",
+        .method = "PUT",
+        .resourceId = "3",
+        .dataset = {{Keys::Role::ID, {CORRECT_ID}},
+                    {Keys::Role::NAME, {CORRECT_NAME}}}};
+
+    std::vector<Role> expectedUsers{};
+    EXPECT_CALL(*rolesRepositoryMock, update(_))
+        .WillOnce(Return());
```
Comment (VolodymyrBilenko) [inline _server/tests/unit/UserModuleTest.cpp]: line shift?

File: _server/tests/component/BusinessLogicTest.cpp [hunk 11]
```diff
@@ -71,6 +72,33 @@ TEST_F(BusinessLogicTest, UsersModule_LoginUser)
...
+    {
+        callbackInvoked = true;
+        ASSERT_TRUE(responseData.dataset.find(Keys::_ERROR) == responseData.dataset.end());
+    };
+
+    // Execute the task
+    businessLogic->executeTask(requestData, callback);
+
+    // Verify the callback was invoked
+    ASSERT_TRUE(callbackInvoked);
```
Comment (VolodymyrBilenko) [inline _server/tests/component/BusinessLogicTest.cpp]: what it the purpose of this check? callbackInvoked will be always true at this part of code, won't it?

**Review by VolodymyrBilenko — CHANGES_REQUESTED** [review]

File: _server/src/BusinessLogic/UsersModule.cpp (hunk 6 above)
Comment (Sage-Cat) [reply _server/src/BusinessLogic/UsersModule.cpp:190]: @VolodymyrBilenko he uses `operator[]` here in the context of inserting new value to map. It's different case from the one with `at()`. Although it's not a best practice, it's still fine. 

We also could consider using something like `std::unordered_map::emplace` that is modern C++20 best practice, but we will need to make a refactoring task for this, cos my code as well as code on client also should be also reviewed to match.

File: _server/src/BusinessLogic/UsersModule.cpp (hunk 7 above)
Comment (Sage-Cat) [reply _server/src/BusinessLogic/UsersModule.cpp:223]: @VolodymyrBilenko as a reference of what?
Nvm, about this, RVO works fine in C++17 and higher. We don't need to worry about this stuff using C++20 compiler.

File: _server/src/BusinessLogic/UsersModule.cpp (hunk 5 above)
Comment (Sage-Cat) [reply _server/src/BusinessLogic/UsersModule.cpp]: @DanyloPlotskykh make it void for my code above also. Anyway you do similar stuff now, so you can cleanup the milk.

File: _server/src/BusinessLogic/UsersModule.hpp (hunk 8 above)
Comment (Sage-Cat) [reply _server/src/BusinessLogic/UsersModule.hpp:27]: @VolodymyrBilenko why?

File: _server/tests/unit/CMakeLists.txt (hunk 9 above)
Comment (Sage-Cat) [reply _server/tests/unit/CMakeLists.txt]: @VolodymyrBilenko he just copied my code I guess. It's fine. It will be a convention for "readability", even if it's kind of weird

File: _server/tests/unit/UserModuleTest.cpp [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+            .WillRepeatedly(Return(usersRepositoryMock));
+        EXPECT_CALL(*repositoryManagerMock, getRolesRepository())
+            .WillRepeatedly(Return(rolesRepositoryMock));
+
+        usersModule = std::make_unique<UsersModule>(*repositoryManagerMock);
+    }
+};
+TEST_F(UserModuleTest, logUser)
+{
+
```
Comment (Sage-Cat) [inline _server/tests/unit/UserModuleTest.cpp:40]: remove space

File: _server/tests/unit/UserModuleTest.cpp [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+          rolesRepositoryMock(std::make_shared<RepositoryMock<Role>>())
+    {
+        EXPECT_CALL(*repositoryManagerMock, getUsersRepository())
+            .WillRepeatedly(Return(usersRepositoryMock));
+        EXPECT_CALL(*repositoryManagerMock, getRolesRepository())
+            .WillRepeatedly(Return(rolesRepositoryMock));
+
+        usersModule = std::make_unique<UsersModule>(*repositoryManagerMock);
+    }
+};
```
Comment (Sage-Cat) [inline _server/tests/unit/UserModuleTest.cpp:39]: add space

File: _server/tests/unit/UserModuleTest.cpp [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+        .method = "POST",
+        .resourceId = "",
+        .dataset = {{Keys::User::USERNAME, {CORRECT_USERNAME}},
+                    {Keys::User::PASSWORD, {CORRECT_PASSWORD}}}};
+    std::vector<User> expectedUsers{User("", CORRECT_USERNAME, CORRECT_PASSWORD, "")};
+    EXPECT_CALL(*usersRepositoryMock, getByField(_, _))
+        .WillOnce(Return(expectedUsers));
+
+    ResponseData response = usersModule->executeTask(requestData);
+    for (size_t i = 0; i < expectedUsers.size(); i++)
```
Comment (Sage-Cat) [inline _server/tests/unit/UserModuleTest.cpp]: It's just a test for only one matched user, so you don't need loop here

File: _server/tests/unit/UserModuleTest.cpp [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+    {
+        EXPECT_CALL(*repositoryManagerMock, getUsersRepository())
+            .WillRepeatedly(Return(usersRepositoryMock));
+        EXPECT_CALL(*repositoryManagerMock, getRolesRepository())
+            .WillRepeatedly(Return(rolesRepositoryMock));
+
+        usersModule = std::make_unique<UsersModule>(*repositoryManagerMock);
+    }
+};
+TEST_F(UserModuleTest, logUser)
```
Comment (Sage-Cat) [inline _server/tests/unit/UserModuleTest.cpp]: name tests same way as named functions = loginUser

File: _server/tests/unit/UserModuleTest.cpp [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+
+    ResponseData response = usersModule->executeTask(requestData);
+    for (size_t i = 0; i < expectedUsers.size(); i++)
+    {
+        EXPECT_EQ(response.dataset.at(Keys::User::PASSWORD).front(), expectedUsers[i].password);
+        EXPECT_EQ(response.dataset.at(Keys::User::USERNAME).front(), expectedUsers[i].username);
+    }
+}
+
+TEST_F(UserModuleTest, regUser)
```
Comment (Sage-Cat) [inline _server/tests/unit/UserModuleTest.cpp]: same

File: _server/tests/unit/UserModuleTest.cpp [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+
+    ResponseData response = usersModule->executeTask(requestData);
+    for (size_t i = 0; i < expectedUsers.size(); i++)
+    {
+        EXPECT_EQ(response.dataset.at(Keys::User::PASSWORD).front(), expectedUsers[i].password);
+        EXPECT_EQ(response.dataset.at(Keys::User::USERNAME).front(), expectedUsers[i].username);
+    }
+}
+
+TEST_F(UserModuleTest, getRoles_Test)
```
Comment (Sage-Cat) [inline _server/tests/unit/UserModuleTest.cpp]: remove _Test from suite name

File: _server/tests/unit/UserModuleTest.cpp [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+        .resourceId = "",
+        .dataset = {{Keys::Role::ID, {CORRECT_ID}},
+                    {Keys::Role::NAME, {CORRECT_NAME}}}};
+
+    std::vector<Role> expectedUsers{Role(CORRECT_ID, CORRECT_NAME)};
+    EXPECT_CALL(*rolesRepositoryMock, add(_))
+        .WillOnce(Return());
+
+    ResponseData response = usersModule->executeTask(requestData);
+    for (size_t i = 0; i < expectedUsers.size(); i++)
```
Comment (Sage-Cat) [inline _server/tests/unit/UserModuleTest.cpp]: remove loop for the suites that test everything except GET requests where is needed more than one data element

File: _server/tests/unit/UserModuleTest.cpp [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+    {
+        EXPECT_EQ(response.dataset.at(Keys::Role::ID).front(), expectedUsers[i].id);
+        EXPECT_EQ(response.dataset.at(Keys::Role::NAME).front(), expectedUsers[i].name);
+    }
+}
+
+TEST_F(UserModuleTest, delete_test)
```
Comment (Sage-Cat) [inline _server/tests/unit/UserModuleTest.cpp]: deleteRole

File: _server/src/BusinessLogic/UsersModule.cpp
Comment (Sage-Cat) [inline _server/src/BusinessLogic/UsersModule.cpp:1]: Attach screenshots of working tests (this tests, not all, in verbose mode) to Jira comments under task.

**Review by Sage-Cat — CHANGES_REQUESTED** [review]

File: _server/tests/unit/CMakeLists.txt (hunk 9 above)
Comment (VolodymyrBilenko) [reply _server/tests/unit/CMakeLists.txt]: Okay, in such case please rename it, since it is not _DIR but _SOURCES

File: _server/src/BusinessLogic/UsersModule.hpp (hunk 8 above)
Comment (VolodymyrBilenko) [reply _server/src/BusinessLogic/UsersModule.hpp:27]: ah, I forgot that it not mandatory for C++. @DanyloPlotskykh You can ignore it.

File: _server/src/BusinessLogic/UsersModule.cpp
Comment (DanyloPlotskykh) [reply _server/src/BusinessLogic/UsersModule.cpp:1]: done

File: _server/src/BusinessLogic/UsersModule.cpp (hunk 1 above)
Comment (DanyloPlotskykh) [reply _server/src/BusinessLogic/UsersModule.cpp]: done

File: _server/src/BusinessLogic/UsersModule.cpp (hunk 3 above)
Comment (DanyloPlotskykh) [reply _server/src/BusinessLogic/UsersModule.cpp]: done

File: _server/src/BusinessLogic/UsersModule.cpp (hunk 4 above)
Comment (DanyloPlotskykh) [reply _server/src/BusinessLogic/UsersModule.cpp]: done

File: _server/src/BusinessLogic/UsersModule.cpp (hunk 2 above)
Comment (DanyloPlotskykh) [reply _server/src/BusinessLogic/UsersModule.cpp]: done

File: _server/src/BusinessLogic/UsersModule.cpp (hunk 5 above)
Comment (DanyloPlotskykh) [reply _server/src/BusinessLogic/UsersModule.cpp]: done

File: _server/tests/unit/UserModuleTest.cpp (hunk 13 above)
Comment (DanyloPlotskykh) [reply _server/tests/unit/UserModuleTest.cpp:39]: done

File: _server/tests/unit/UserModuleTest.cpp (hunk 15 above)
Comment (DanyloPlotskykh) [reply _server/tests/unit/UserModuleTest.cpp]: done

File: _server/tests/unit/UserModuleTest.cpp (hunk 12 above)
Comment (DanyloPlotskykh) [reply _server/tests/unit/UserModuleTest.cpp:40]: done

File: _server/tests/unit/UserModuleTest.cpp (hunk 14 above)
Comment (DanyloPlotskykh) [reply _server/tests/unit/UserModuleTest.cpp]: done

File: _server/tests/unit/UserModuleTest.cpp (hunk 16 above)
Comment (DanyloPlotskykh) [reply _server/tests/unit/UserModuleTest.cpp]: done

File: _server/tests/unit/UserModuleTest.cpp (hunk 17 above)
Comment (DanyloPlotskykh) [reply _server/tests/unit/UserModuleTest.cpp]: done

File: _server/tests/unit/UserModuleTest.cpp (hunk 18 above)
Comment (DanyloPlotskykh) [reply _server/tests/unit/UserModuleTest.cpp]: done

File: _server/tests/unit/UserModuleTest.cpp (hunk 19 above)
Comment (DanyloPlotskykh) [reply _server/tests/unit/UserModuleTest.cpp]: done

File: _server/src/BusinessLogic/UsersModule.cpp
Comment (Sage-Cat) [reply _server/src/BusinessLogic/UsersModule.cpp:1]: @DanyloPlotskykh attach more detailed view for logs to Jira ticket

File: _server/tests/unit/CMakeLists.txt (hunk 9 above)
Comment (Sage-Cat) [reply _server/tests/unit/CMakeLists.txt]: @VolodymyrBilenko oh, I've missed that.

@DanyloPlotskykh fix this.

**Review by Sage-Cat — CHANGES_REQUESTED** [review]

File: _server/tests/unit/UserModuleTest.cpp [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+    constexpr char CORRECT_NAME[] = "admin3";
+    RequestData requestData{
+        .module = "users",
+        .submodule = "roles",
+        .method = "PUT",
+        .resourceId = "3",
+        .dataset = {{Keys::Role::ID, {CORRECT_ID}},
+                    {Keys::Role::NAME, {CORRECT_NAME}}}};
+
+    std::vector<Role> expectedUsers{};
```
Comment (Sage-Cat) [inline _server/tests/unit/UserModuleTest.cpp]: What does this test? Nothing?

File: _server/tests/unit/UserModuleTest.cpp [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+    constexpr char CORRECT_NAME[] = "admin4";
+    RequestData requestData{
+        .module = "users",
+        .submodule = "roles",
+        .method = "DEL",
+        .resourceId = "4",
+        .dataset = {{Keys::Role::ID, {CORRECT_ID}},
+                    {Keys::Role::NAME, {CORRECT_NAME}}}};
+
+    std::vector<Role> expectedUsers{};
```
Comment (Sage-Cat) [inline _server/tests/unit/UserModuleTest.cpp]: Same... Test does nothing

**Review by Sage-Cat — CHANGES_REQUESTED** [review]

File: _server/tests/unit/UserModuleTest.cpp [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+    constexpr char CORRECT_NAME[] = "adminchik";
+    RequestData requestData{
+        .module = "users",
+        .submodule = "roles",
+        .method = "POST",
+        .resourceId = "",
+        .dataset = {{Keys::Role::ID, {CORRECT_ID}},
+                    {Keys::Role::NAME, {CORRECT_NAME}}}};
+
+    std::vector<Role> expectedUsers{Role(CORRECT_ID, CORRECT_NAME)};
```
Comment (Sage-Cat) [inline _server/tests/unit/UserModuleTest.cpp]: expected roles?

**Review by Sage-Cat — CHANGES_REQUESTED** [review]

File: _server/src/BusinessLogic/UsersModule.cpp
Comment (DanyloPlotskykh) [reply _server/src/BusinessLogic/UsersModule.cpp:1]: done

File: _server/tests/unit/CMakeLists.txt (hunk 9 above)
Comment (DanyloPlotskykh) [reply _server/tests/unit/CMakeLists.txt]: done

File: _server/tests/unit/UserModuleTest.cpp (hunk 22 above)
Comment (DanyloPlotskykh) [reply _server/tests/unit/UserModuleTest.cpp]: done

File: _server/tests/unit/UserModuleTest.cpp (hunk 20 above)
Comment (DanyloPlotskykh) [reply _server/tests/unit/UserModuleTest.cpp]: done

File: _server/tests/unit/UserModuleTest.cpp (hunk 21 above)
Comment (DanyloPlotskykh) [reply _server/tests/unit/UserModuleTest.cpp]: done

**Comment by DanyloPlotskykh** [comment]
р

File: _server/tests/component/BusinessLogicTest.cpp (hunk 11 above)
Comment (DanyloPlotskykh) [reply _server/tests/component/BusinessLogicTest.cpp]: done

File: _server/tests/unit/UserModuleTest.cpp (hunk 10 above)
Comment (DanyloPlotskykh) [reply _server/tests/unit/UserModuleTest.cpp]: done

File: _server/src/BusinessLogic/UsersModule.cpp (hunk 4 above)
Comment (VolodymyrBilenko) [reply _server/src/BusinessLogic/UsersModule.cpp]: You didn't change it. Grammatical error is still present. **Array** is a singular, not plural

File: _server/src/BusinessLogic/UsersModule.cpp [hunk 23]
```diff
@@ -123,3 +143,92 @@ ResponseData UsersModule::registerUser(const Dataset &request)
...
+{
+    SPDLOG_TRACE("UsersModule::getRole");
+    ResponseData response;
+    auto rolesVec = m_rolesRepository->getAll();
+
+    if (rolesVec.empty())
+    {
+        response.dataset[Keys::Role::ID] = {""};
+        response.dataset[Keys::Role::NAME] = {""};
+        SPDLOG_WARN("UsersModule::getRoles | array of roles are empty");
```
Comment (VolodymyrBilenko) [inline _server/src/BusinessLogic/UsersModule.cpp]: is empty

File: _server/src/BusinessLogic/UsersModule.cpp [hunk 24]
```diff
@@ -123,3 +139,92 @@ ResponseData UsersModule::registerUser(const Dataset &request)
...
+    catch (const std::out_of_range &e)
+    {
+        SPDLOG_WARN("UsersModule::addNewRole username or password does not exit");
+        throw ServerException(_M, "Server got no username or password in Dataset");
+    }
+    auto usersVec = m_usersRepository->getByField(Keys::Role::NAME, roleName);
+    if (usersVec.empty())
+    {
+        m_rolesRepository->add(Role("", roleName));
+        SPDLOG_INFO("UsersModule::registerUser | new role `{}` is add", roleName);
```
Comment (VolodymyrBilenko) [inline _server/src/BusinessLogic/UsersModule.cpp]: is added

**Review by VolodymyrBilenko — DISMISSED** [review]

File: _server/src/BusinessLogic/UsersModule.cpp (hunk 24 above)
Comment (DanyloPlotskykh) [reply _server/src/BusinessLogic/UsersModule.cpp]: done

File: _server/src/BusinessLogic/UsersModule.cpp (hunk 23 above)
Comment (DanyloPlotskykh) [reply _server/src/BusinessLogic/UsersModule.cpp]: done

File: _server/src/BusinessLogic/UsersModule.cpp [hunk 25]
```diff
@@ -67,7 +86,12 @@ ResponseData UsersModule::loginUser(const Dataset &request)
     // get user by username from repository
     auto usersVec = m_usersRepository->getByField(Keys::User::USERNAME, clientUsername);
 
-    if (!usersVec.empty())
+    if (usersVec.empty())
```
Comment (Sage-Cat) [inline _server/src/BusinessLogic/UsersModule.cpp:89]: Why error goes first?

File: _server/tests/CMakeLists.txt [hunk 26]
```diff
@@ -1,8 +1,9 @@
 find_package(GTest REQUIRED)
 
-set(SERVER_UNIT_TESTS_DIR "${SERVER_TESTS_DIR}/unit")
-set(SERVER_COMPONENT_TESTS_DIR "${SERVER_TESTS_DIR}/component")
-set(SERVER_TEST_MOCKS_DIR "${SERVER_TESTS_DIR}/mocks")
-set(SERVER_TEST_WRAPPERS_DIR "${SERVER_TESTS_DIR}/wrappers")
+set(SERVER_UNIT_TESTS "${SERVER_TESTS_DIR}/unit")
```
Comment (Sage-Cat) [inline _server/tests/CMakeLists.txt]: Why? 
Revert this changes back
It's DIR

File: _server/src/BusinessLogic/UsersModule.cpp (hunk 25 above)
Comment (DanyloPlotskykh) [reply _server/src/BusinessLogic/UsersModule.cpp:89]: image

Comment (DanyloPlotskykh) [reply _server/src/BusinessLogic/UsersModule.cpp:89]: тепер не знаю як краще

File: _server/tests/CMakeLists.txt (hunk 26 above)
Comment (DanyloPlotskykh) [reply _server/tests/CMakeLists.txt]: done

File: _server/src/BusinessLogic/UsersModule.cpp [hunk 27]
```diff
@@ -123,3 +143,92 @@ ResponseData UsersModule::registerUser(const Dataset &request)
...
+    }
+    catch (const std::out_of_range &e)
+    {
+        SPDLOG_WARN("UsersModule::addNewRole username or password does not exit");
+        throw ServerException(_M, "Server got no username or password in Dataset");
+    }
+    auto usersVec = m_rolesRepository->getByField(Keys::Role::NAME, roleName);
+    if (usersVec.empty())
+    {
+        m_rolesRepository->add(Role("", roleName));
```
Comment (Sage-Cat) [inline _server/src/BusinessLogic/UsersModule.cpp]: wtf is this?
`add(Role`

**Review by Sage-Cat — CHANGES_REQUESTED** [review]

File: _server/src/BusinessLogic/UsersModule.cpp (hunk 27 above)
Comment (DanyloPlotskykh) [reply _server/src/BusinessLogic/UsersModule.cpp]: done

**Review by Sage-Cat — APPROVED** [review]

**Review by MaksGMaks — APPROVED** [review]
